File: bench/strwalk.js
// The JavaScript twin of strwalk.sl.
//
// **A JavaScript string is UTF-16 and `s[i]` reads one CODE UNIT in constant time.** Every character
// in this text is in the basic plane, so one unit is one character and the walk below counts exactly
// what the slate program counts -- which is why the text was chosen that way. An emoji would break
// the correspondence: it is one character in slate and two units here, and `s.length` would be a
// different number.
//
// **So this twin does less work than the slate program by construction**, storing two bytes per
// character rather than a variable one to four, and `s.length` is a field rather than anything that
// has to be counted.

function run(s) {
    let found = 0;
    let i = 0;

    while (i < s.length) {
        if (s[i] === "本") {
            found = found + 1;
        }

        i = i + 1;
    }

    return found;
}

const parts = [];

for (let k = 0; k < 1000; k++) {
    parts.push("日本語あいうえおかきく" + "さしすせそたちつてと");
}

console.log(run(parts.join("")));

File: bench/strwalk.py
# The Python twin of strwalk.sl.
#
# **`s[i]` is a constant-time read of a CODE POINT**, which is the same unit slate indexes -- so of
# the three twins this is the one whose walk means exactly what the slate program's means.
#
# **It is constant time because CPython picks a fixed width per string** (PEP 393): a string holding
# any character above U+00FF is stored four bytes per code point, so this text costs four times its
# UTF-8 size in memory and an index is a multiplication. That is the trade slate does not make -- it
# keeps the UTF-8 and remembers where it last was.


def run(s):
    found, i = 0, 0

    while i < len(s):
        if s[i] == "本":
            found = found + 1

        i = i + 1

    return found


print(run("".join("日本語あいうえおかきく" + "さしすせそたちつてと" for _ in range(1000))))
